doc/man/innbind.8
.\" Automatically generated by Pod::Man v1.37, Pod::Parser v1.14
.\"
.\" Standard preamble:
.\" ========================================================================
.de Sh \" Subsection heading
.br
.if t .Sp
.ne 5
.PP
\fB\\$1\fR
.PP
..
.de Sp \" Vertical space (when we can't use .PP)
.if t .sp .5v
.if n .sp
..
.de Vb \" Begin verbatim text
.ft CW
.nf
.ne \\$1
..
.de Ve \" End verbatim text
.ft R
.fi
..
.\" Set up some character translations and predefined strings.  \*(-- will
.\" give an unbreakable dash, \*(PI will give pi, \*(L" will give a left
.\" double quote, and \*(R" will give a right double quote.  | will give a
.\" real vertical bar.  \*(C+ will give a nicer C++.  Capital omega is used to
.\" do unbreakable dashes and therefore won't be available.  \*(C` and \*(C'
.\" expand to `' in nroff, nothing in troff, for use with C<>.
.tr \(*W-|\(bv\*(Tr
.ds C+ C\v'-.1v'\h'-1p'\s-2+\h'-1p'+\s0\v'.1v'\h'-1p'
.ie n \{\
.    ds -- \(*W-
.    ds PI pi
.    if (\n(.H=4u)&(1m=24u) .ds -- \(*W\h'-12u'\(*W\h'-12u'-\" diablo 10 pitch
.    if (\n(.H=4u)&(1m=20u) .ds -- \(*W\h'-12u'\(*W\h'-8u'-\"  diablo 12 pitch
.    ds L" ""
.    ds R" ""
.    ds C` ""
.    ds C' ""
'br\}
.el\{\
.    ds -- \|\(em\|
.    ds PI \(*p
.    ds L" ``
.    ds R" ''
'br\}
.\"
.\" If the F register is turned on, we'll generate index entries on stderr for
.\" titles (.TH), headers (.SH), subsections (.Sh), items (.Ip), and index
.\" entries marked with X<> in POD.  Of course, you'll have to process the
.\" output yourself in some meaningful fashion.
.if \nF \{\
.    de IX
.    tm Index:\\$1\t\\n%\t"\\$2"
..
.    nr % 0
.    rr F
.\}
.\"
.\" For nroff, turn off justification.  Always turn off hyphenation; it makes
.\" way too many mistakes in technical documents.
.hy 0
.if n .na
.\"
.\" Accent mark definitions (@(#)ms.acc 1.5 88/02/08 SMI; from UCB 4.2).
.\" Fear.  Run.  Save yourself.  No user-serviceable parts.
.    \" fudge factors for nroff and troff
.if n \{\
.    ds #H 0
.    ds #V .8m
.    ds #F .3m
.    ds #[ \f1
.    ds #] \fP
.\}
.if t \{\
.    ds #H ((1u-(\\\\n(.fu%2u))*.13m)
.    ds #V .6m
.    ds #F 0
.    ds #[ \&
.    ds #] \&
.\}
.    \" simple accents for nroff and troff
.if n \{\
.    ds ' \&
.    ds ` \&
.    ds ^ \&
.    ds , \&
.    ds ~ ~
.    ds /
.\}
.if t \{\
.    ds ' \\k:\h'-(\\n(.wu*8/10-\*(#H)'\'\h"|\\n:u"
.    ds ` \\k:\h'-(\\n(.wu*8/10-\*(#H)'\`\h'|\\n:u'
.    ds ^ \\k:\h'-(\\n(.wu*10/11-\*(#H)'^\h'|\\n:u'
.    ds , \\k:\h'-(\\n(.wu*8/10)',\h'|\\n:u'
.    ds ~ \\k:\h'-(\\n(.wu-\*(#H-.1m)'~\h'|\\n:u'
.    ds / \\k:\h'-(\\n(.wu*8/10-\*(#H)'\z\(sl\h'|\\n:u'
.\}
.    \" troff and (daisy-wheel) nroff accents
.ds : \\k:\h'-(\\n(.wu*8/10-\*(#H+.1m+\*(#F)'\v'-\*(#V'\z.\h'.2m+\*(#F'.\h'|\\n:u'\v'\*(#V'
.ds 8 \h'\*(#H'\(*b\h'-\*(#H'
.ds o \\k:\h'-(\\n(.wu+\w'\(de'u-\*(#H)/2u'\v'-.3n'\*(#[\z\(de\v'.3n'\h'|\\n:u'\*(#]
.ds d- \h'\*(#H'\(pd\h'-\w'~'u'\v'-.25m'\f2\(hy\fP\v'.25m'\h'-\*(#H'
.ds D- D\\k:\h'-\w'D'u'\v'-.11m'\z\(hy\v'.11m'\h'|\\n:u'
.ds th \*(#[\v'.3m'\s+1I\s-1\v'-.3m'\h'-(\w'I'u*2/3)'\s-1o\s+1\*(#]
.ds Th \*(#[\s+2I\s-2\h'-\w'I'u*3/5'\v'-.3m'o\v'.3m'\*(#]
.ds ae a\h'-(\w'a'u*4/10)'e
.ds Ae A\h'-(\w'A'u*4/10)'E
.    \" corrections for vroff
.if v .ds ~ \\k:\h'-(\\n(.wu*9/10-\*(#H)'\s-2\u~\d\s+2\h'|\\n:u'
.if v .ds ^ \\k:\h'-(\\n(.wu*10/11-\*(#H)'\v'-.4m'^\v'.4m'\h'|\\n:u'
.    \" for low resolution devices (crt and lpr)
.if \n(.H>23 .if \n(.V>19 \
\{\
.    ds : e
.    ds 8 ss
.    ds o a
.    ds d- d\h'-1'\(ga
.    ds D- D\h'-1'\(hy
.    ds th \o'bp'
.    ds Th \o'LP'
.    ds ae ae
.    ds Ae AE
.\}
.rm #[ #] #H #V #F C
.\" ========================================================================
.\"
.IX Title "INNBIND 8"
.TH INNBIND 8 "2004-05-19" "INN 2.5.0" "InterNetNews Documentation"
.SH "NAME"
innbind \- Helper program to bind sockets to privileged ports
.SH "SYNOPSIS"
.IX Header "SYNOPSIS"
\&\fBinnbind\fR [\fB\-p\fR] \fIfd\fR,\fIfamily\fR,\fIaddress\fR,\fIport\fR [...]
.SH "DESCRIPTION"
.IX Header "DESCRIPTION"
\&\fBinnbind\fR is a helper program that's not meant to be run directly.
Instead, \fBinnd\fR and \fBnnrpd\fR use it internally to bind to ports that
require root privileges to bind to.
.PP
This program must be installed setuid root in order for \fBinnd\fR or
\&\fBnnrpd\fR to bind to ports under 1024.  The only functions that it's
willing to perform are to bind an open file descriptor to a given address
and port or to create a new socket, bind it, and return the bound socket
to its caller.  It can only be run as the news user (as specified at
configure time), and will only bind to ports 119, 433, an additional port
specified with the \fB\-\-with\-innd\-port\fR argument to configure, or ports
numbered 1024 or higher.
.PP
Each argument to \fBinnbind\fR must be a comma-separated list of four
elements.  The first is the file descriptor number that should be bound,
the second is the numeric family of the socket (\s-1AF_INET\s0 or \s-1AF_INET6\s0), the
third is the local address to bind to (in dotted quad format for IPv4 and
in colon-separated address format for IPv6), and the fourth is the port
number.  To bind to all addresses with IPv4, use \f(CW0.0.0.0\fR as the
address.  To bind to all addresses with IPv6, use \f(CW\*(C`::\*(C'\fR as the address.
.PP
Multiple arguments can be specified to tell \fBinnbind\fR to bind multiple
sockets at the same time.  Any errors (other than permission denied \*(-- see
below) encountered will cause \fBinnbind\fR to abort, and error messages will
be sent both to syslog and to standard error.
.PP
By default, \fBinnbind\fR attempts to just bind the already open file
descriptor that it inherits from its caller.  For each successfully bound
file descriptor (in the order given on the command line), \fBinnbind\fR
prints \f(CW\*(C`ok\*(C'\fR and a newline to its standard output.
.PP
On some systems (apparently just STREAMS-based systems), however, even a
setuid root program cannot bind a socket to a privileged port that was
created by a process that didn't have permissions.  If \fBinnbind\fR gets
permission denied when trying to bind a socket, it will print \f(CW\*(C`no\*(C'\fR and a
newline to its standard output.  It will then create a new socket, bind it
as specified, and then attempt to pass that socket back to its caller
using the I_SENDFD \s-1STREAMS\s0 ioctl.  The caller should receive that file
descriptor with I_RECVFD and use it instead of the one that it created.
.PP
Note that file descriptor passing is only supported on STREAMS-based
systems since it is done with ioctls over a pipe.  However, it is believed
that those systems are exactly the systems that can't simply bind the
inherited file descriptor.  If this assumption proves to be incorrect,
traditional \s-1BSD\s0 file descriptor passing over a Unix domain socket will
have to be added.
.SH "OPTIONS"
.IX Header "OPTIONS"
.IP "\fB\-p\fR" 4
.IX Item "-p"
If given as the first command-line argument, no attempt will be made to
bind the inherited file descriptor and \fBinnbind\fR will only try creation
of a new file descriptor and passing it back via standard output.  This
option is primarily useful for testing.
.SH "SECURITY"
.IX Header "SECURITY"
As \fBinnbind\fR is normally installed setuid root, security is even more of
an issue for it than for other parts of \s-1INN\s0.  It is a fairly short
program, and if you understand C, you are encouraged to audit it yourself
to be certain that it does only what it is supposed to do.  The only \s-1INN\s0
library functions it uses are the vector functions, the message functions
for error reporting, and xstrdup.
.PP
The ports that will be bound are restricted to prevent potential attacks
made possible by the ability to bind low-numbered ports, such as exploits
of the \fIrsh\fR\|(1) family of commands on some systems.  If \fBinnbind\fR is
installed setuid root, it can only be executed by the news user to prevent
other users on the system from being able to bind to even those few
privileged ports that it allows.
.PP
\&\fBinnbind\fR uses no external configuration files; the only files it might
open are through the system \fIgetpwnam\fR\|(3) service to get the \s-1UID\s0 of the news
user.  The only user input that it accepts are its command-line arguments.
.SH "DIAGNOSTICS"
.IX Header "DIAGNOSTICS"
\&\fBinnbind\fR may log the following messages to syslog and print them to
standard error.
.ie n .IP "cannot create socket for %s:\fR \f(CW%s" 4
.el .IP "cannot create socket for \f(CW%s:\fR \f(CW%s\fR" 4
.IX Item "cannot create socket for %s: %s"
(Fatal) \fBinnbind\fR fell back on attempting to create a new socket to bind
for the given argument, and the socket creation failed.
.ie n .IP "cannot bind socket for %s:\fR \f(CW%s" 4
.el .IP "cannot bind socket for \f(CW%s:\fR \f(CW%s\fR" 4
.IX Item "cannot bind socket for %s: %s"
(Fatal) Calling bind for the socket corresponding to the given argument
failed with a system error.  If the error indicates permission denied,
make sure that \fBinnbind\fR is setuid root.  This can also be caused by
trying to use IPv6 on a system whose kernel does not support it.
.ie n .IP "cannot bind to restricted port %hu\fR in \f(CW%s" 4
.el .IP "cannot bind to restricted port \f(CW%hu\fR in \f(CW%s\fR" 4
.IX Item "cannot bind to restricted port %hu in %s"
(Fatal) The port number portion of the given command-line argument is for
a port below 1024 which is not 119, 433, or a port given to
\&\fB\-\-with\-innd\-port\fR at configure time.  Other ports are not allowed for
security reasons.
.ie n .IP "cannot get socket options for file descriptor %d:\fR \f(CW%s" 4
.el .IP "cannot get socket options for file descriptor \f(CW%d:\fR \f(CW%s\fR" 4
.IX Item "cannot get socket options for file descriptor %d: %s"
(Fatal) \fBinnbind\fR was unable to get the socket options for that file
descriptor.  The most likely cause of this error is passing the wrong file
descriptor number to \fBinnbind\fR (a file descriptor that isn't open, or
that corresponds to a regular file rather than a network socket).
.ie n .IP "cannot get \s-1UID\s0 for %s" 4
.el .IP "cannot get \s-1UID\s0 for \f(CW%s\fR" 4
.IX Item "cannot get UID for %s"
(Fatal) \fBinnbind\fR was unable to get the \s-1UID\s0 for the news user specified
during configure (and defaulting to \f(CW\*(C`news\*(C'\fR).  This normally means that
user isn't in the system passwd file.
.ie n .IP "cannot mark socket reusable for %s:\fR \f(CW%s" 4
.el .IP "cannot mark socket reusable for \f(CW%s:\fR \f(CW%s\fR" 4
.IX Item "cannot mark socket reusable for %s: %s"
(Fatal) \fBinnbind\fR created a new socket for the given argument but was
unable to mark its bind address reusable (the \s-1SO_REUSEADDR\s0 socket option).
.ie n .IP "cannot pass file descriptor: %s" 4
.el .IP "cannot pass file descriptor: \f(CW%s\fR" 4
.IX Item "cannot pass file descriptor: %s"
(Fatal) \fBinnbind\fR created and bound a new file descriptor but was unable
to pass it back to its caller via its standard output, using the I_SENDFD
\&\s-1STREAMS\s0 ioctl.
.ie n .IP "invalid file descriptor %d: not \s-1SOCK_STREAM\s0" 4
.el .IP "invalid file descriptor \f(CW%d:\fR not \s-1SOCK_STREAM\s0" 4
.IX Item "invalid file descriptor %d: not SOCK_STREAM"
(Fatal) The given file descriptor is not a \s-1SOCK_STREAM\s0 socket.  \fBinnbind\fR
can only bind \s-1SOCK_STREAM\s0 sockets.
.ie n .IP "invalid IPv4 address %s\fR in \f(CW%s" 4
.el .IP "invalid IPv4 address \f(CW%s\fR in \f(CW%s\fR" 4
.IX Item "invalid IPv4 address %s in %s"
(Fatal) The IPv4 address specified in the given command-line option could
not be parsed by \fIinet_aton\fR\|(3).  IPv4 addresses should be specified in the
standard dotted quad format (10.2.3.4).
.ie n .IP "invalid IPv6 address %s\fR in \f(CW%s" 4
.el .IP "invalid IPv6 address \f(CW%s\fR in \f(CW%s\fR" 4
.IX Item "invalid IPv6 address %s in %s"
(Fatal) The IPv6 address specified in the given command-line option could
not be parsed by \fIinet_pton\fR\|(3).  IPv6 addresses should be specified in \s-1RFC\s0
2373 format (1080:0:0:0:8:800:200C:417A or 1080::8:800:200C:417A).
.ie n .IP "invalid command-line argument %s" 4
.el .IP "invalid command-line argument \f(CW%s\fR" 4
.IX Item "invalid command-line argument %s"
(Fatal) The specified command-line argument could not be parsed or was not
in the correct format.
.ie n .IP "invalid file descriptor %s\fR in \f(CW%s" 4
.el .IP "invalid file descriptor \f(CW%s\fR in \f(CW%s\fR" 4
.IX Item "invalid file descriptor %s in %s"
(Fatal) The file descriptor portion of the given command-line argument is
not a non-negative integer.
.ie n .IP "invalid port number %s\fR in \f(CW%s" 4
.el .IP "invalid port number \f(CW%s\fR in \f(CW%s\fR" 4
.IX Item "invalid port number %s in %s"
(Fatal) The port number portion of the given command-line argument is not
a non-negative integer.
.ie n .IP "invalid protocol family %s\fR in \f(CW%s" 4
.el .IP "invalid protocol family \f(CW%s\fR in \f(CW%s\fR" 4
.IX Item "invalid protocol family %s in %s"
(Fatal) The protocol family portion of the given command-line argument is
not a non-negative integer.  It should be equal to either \s-1AF_INET\s0 or
\&\s-1AF_INET6\s0 on the system where \fBinnbind\fR is run.
.ie n .IP "must be run by user %s\fR (%lu), not \f(CW%lu" 4
.el .IP "must be run by user \f(CW%s\fR (%lu), not \f(CW%lu\fR" 4
.IX Item "must be run by user %s (%lu), not %lu"
(Fatal) When setuid root, \fBinnbind\fR may only be run by the news user as
specified at configure time (\f(CW\*(C`news\*(C'\fR by default), for security reasons.
.IP "no addresses specified" 4
.IX Item "no addresses specified"
(Fatal) No arguments were given on the command line (except maybe \fB\-p\fR).
.ie n .IP "port may not be zero in %s" 4
.el .IP "port may not be zero in \f(CW%s\fR" 4
.IX Item "port may not be zero in %s"
(Fatal) The port number portion of the given command-line argument was
zero.
.ie n .IP "unknown protocol family %s\fR in \f(CW%s" 4
.el .IP "unknown protocol family \f(CW%s\fR in \f(CW%s\fR" 4
.IX Item "unknown protocol family %s in %s"
(Fatal) The protocol number portion of the given command-line argument is
neither \s-1AF_INET\s0 nor \s-1AF_INET6\s0.
.SH "EXAMPLES"
.IX Header "EXAMPLES"
As mentioned above, \fBinnbind\fR is never run directly, only by \fBinnd\fR and
other programs that need to bind to and listen to network ports.  Sample
invocations by \fBinnd\fR would be:
.PP
.Vb 1
\&    innbind 3,10,::,119
.Ve
.PP
to bind the IPv6 socket on file descriptor 3 to port 119, all addresses,
or:
.PP
.Vb 1
\&    innbind 6,2,10.0.0.3,433
.Ve
.PP
to bind the IPv4 socket on file descriptor 6 to port 433 in the address
10.0.0.3.
.SH "HISTORY"
.IX Header "HISTORY"
Written by Russ Allbery <rra@stanford.edu> for InterNetNews.
.PP
$Id$
.SH "SEE ALSO"
.IX Header "SEE ALSO"
\&\fIinet_aton\fR\|(3), \fIinet_pton\fR\|(3), \fIinnd\fR\|(8), \fInnrpd\fR\|(8)


NEWS
Current

    Here are the changes for the latest version of INN.

    * New readers.conf file (replaces nnrp.access) which allows more
      flexible specification of access restrictions.

    * Unified overview has been replaced with an overview API, and there are
      now two separate overview implementations to choose from.  One
      (tradindexed) is very like traditional overview but uses an additional
      index file for faster responses to readers.  The other (buffindexed)
      uses circular buffers and can handle a higher incoming article rate
      while still being fast for readers.  The ovmethod key in inn.conf
      chooses the overview method to use.

    * All article storage and retrieval is now done via the storage API. 
      Traditional spool is now available as a storage type under the storage
      API.

    * The timecaf storage method has been added, similar to timehash but
      storing multiple articles in a single file.  See INSTALL for details
      on it.

    * INN now supports embedded Python filters as well as Perl and TCL
      filters.

Version 2.2

    * New storage.conf file (replaces storage.ctl).

    * New (optional) way of handling non-cancel control messages
      (controlchan) that serializes them and prevents server overload from
      control message storms.

    * Support for actsyncd to fetch active file with ftp; configured by
      default to get the one at ftp://ftp.isc.org/pub/usenet/CONFIG/active.Z
      if you run actsyncd.  Be sure to read the manual page for actsync to
      configure an actsync.ign file for your site, and test simpleftp if you
      do not configure with wget or ncftp.  Also see
      ftp://ftp.isc.org/pub/usenet/CONFIG/README.

    * Some options to configure are now moved to inn.conf (merge-to-groups
      and pgp-verify).

    * inndf, a portable version of df(1) is supplied.

    * New cnfsstat program to show stats of cnfs buffers.

    * news2mail program for gateway'ing news into email is provided.

    * mailpost program for gateway'ing email to news is provided.

    * pullnews program for doing a sucking feed is provided (not meant for
      large feeds).

    * The innshellvars.csh.in script is obsolete (and lives in the obsolete
      directory, for now).



backends/innbind.c
/*
**  Helper program to bind a socket to a low-numbered port.
**
**  Written by Russ Allbery <eagle@eyrie.org>
*/

#include "portable/system.h"

#include "portable/socket.h"
#include <errno.h>
#ifdef HAVE_STREAMS_SENDFD
#    include <stropts.h>
#endif
#include <syslog.h>

#include "inn/libinn.h"
#include "inn/messages.h"
#include "inn/newsuser.h"
#include "inn/vector.h"

/* Macros to set the len attribute of sockaddrs. */
#if HAVE_STRUCT_SOCKADDR_SA_LEN
#    define sin_set_length(s)  ((s)->sin_len = sizeof(struct sockaddr_in))
#    define sin6_set_length(s) ((s)->sin6_len = sizeof(struct sockaddr_in6))
#else
#    define sin_set_length(s)  /* empty */
#    define sin6_set_length(s) /* empty */
#endif

/* INND_PORT is the additional port specified at configure time to which the
   news user should be allowed to bind.  If it's not set, set it to 119 (which
   will cause it to have no effect).  I hate #ifdef in code, can you tell? */
#ifndef INND_PORT
#    define INND_PORT 119
#endif

/* Holds the information about a network socket to bind. */
struct binding {
    int fd;
    int family;
    char *address;
    unsigned short port;
};


/*
**  Convert a string to a number with error checking, returning true if the
**  number was parsed correctly and false otherwise.  Stores the converted
**  number in the second argument.  Equivalent to calling strtol, but with the
**  base always fixed at 10, with checking of errno, ensuring that all of the
**  string is consumed, and checking that the resulting number is positive.
*/
static bool
convert_string(const char *string, long *result)
{
    char *end;

    if (*string == '\0')
        return false;
    errno = 0;
    *result = strtol(string, &end, 10);
    if (errno != 0 || *end != '\0' || *result < 0)
        return false;
    return true;
}


/*
**  Parse a command-line argument into a struct binding.  The command line
**  argument is four comma-separated values: the file descriptor, the family,
**  the listening address, and the port number.  The caller is responsible for
**  freeing the address attribute of the supplied binding struct, although if
**  a binding struct is passed in for use and has a non-NULL address, it will
**  be freed first.
*/
static void
parse_argument(const char *string, struct binding *binding)
{
    struct vector *spec;
    long value;

    /* Do the initial parse and allocate our data structures. */
    spec = vector_split(string, ',', NULL);
    if (spec->count != 4)
        die("invalid command-line argument %s", string);

    /* Get the file descriptor, address family, and port. */
    if (!convert_string(spec->strings[0], &value))
        die("invalid file descriptor %s in %s", spec->strings[0], string);
    binding->fd = value;
    if (!convert_string(spec->strings[1], &value))
        die("invalid protocol family %s in %s", spec->strings[1], string);
    binding->family = value;
    if (binding->address != NULL)
        free(binding->address);
    binding->address = xstrdup(spec->strings[2]);
    if (!convert_string(spec->strings[3], &value))
        die("invalid port number %s in %s", spec->strings[3], string);
    if (value == 0)
        die("port may not be zero in %s", string);
    binding->port = value;

    /* Done.  Clean up. */
    vector_free(spec);
}


/*
**  Bind an IPv4 address, given the file descriptor, string giving the
**  address, and the port.  The fourth argument is the full binding
**  specification for error reporting.  Returns true on success, false if
**  binding failed due to permission denied.  Die on any other failure.
*/
static bool
bind_ipv4(int fd, const char *address, unsigned short port, const char *spec)
{
    struct sockaddr_in server;
    struct in_addr addr;

    memset(&server, '\0', sizeof(server));
    server.sin_family = AF_INET;
    server.sin_port = htons(port);
    if (!inet_aton(address, &addr))
        die("invalid IPv4 address %s in %s", address, spec);
    server.sin_addr = addr;
    sin_set_length(&server);
    if (bind(fd, (struct sockaddr *) &server, sizeof(server)) < 0) {
        if (errno == EACCES)
            return false;
        else
            sysdie("cannot bind socket for %s", spec);
    }
    return true;
}


/*
**  Bind an IPv6 address, given the file descriptor, string giving the
**  address, and the port.  The fourth argument is the full binding
**  specification for error reporting.  Returns true on success, false if
**  binding failed due to permission denied.  Die on any other failure.
*/
#ifdef HAVE_INET6
static bool
bind_ipv6(int fd, const char *address, unsigned short port, const char *spec)
{
    struct sockaddr_in6 server;
    struct in6_addr addr;

    memset(&server, '\0', sizeof(server));
    server.sin6_family = AF_INET6;
    server.sin6_port = htons(port);
    if (inet_pton(AF_INET6, address, &addr) < 1)
        die("invalid IPv6 address %s in %s", address, spec);
    server.sin6_addr = addr;
    sin6_set_length(&server);
    if (bind(fd, (struct sockaddr *) &server, sizeof(server)) < 0) {
        if (errno == EACCES)
            return false;
        else
            sysdie("cannot bind socket for %s", spec);
    }
    return true;
}
#endif /* HAVE_INET6 */


/*
**  Given a struct binding, bind that file descriptor.  Also takes the
**  command-line argument for error reporting.  Returns true on success, false
**  if binding failed due to permission denied.  Die on any other failure.
*/
static bool
bind_address(struct binding *binding, const char *spec)
{
    int fd = binding->fd;
    unsigned short port = binding->port;
    int type;
    socklen_t length;

    /* Make sure that we're allowed to bind to that port. */
    if (port < 1024 && port != 119 && port != 433 && port != 563
        && port != INND_PORT)
        die("cannot bind to restricted port %hu in %s", port, spec);

    /* Sanity check on the socket. */
    length = sizeof(type);
    if (getsockopt(fd, SOL_SOCKET, SO_TYPE, &type, &length) < 0)
        sysdie("cannot get socket options for file descriptor %d", fd);
    if (type != SOCK_STREAM)
        die("invalid file descriptor %d: not SOCK_STREAM", fd);

    /* Based on the address family, parse either an IPv4 or IPv6 address. */
    if (binding->family == AF_INET)
        return bind_ipv4(fd, binding->address, port, spec);
#ifdef HAVE_INET6
    else if (binding->family == AF_INET6)
        return bind_ipv6(fd, binding->address, port, spec);
#endif
    else
        die("unknown protocol family %d in %s", binding->family, spec);
}


/*
**  Given a struct binding, create a socket for it and fill in the file
**  descriptor.  Also takes the command-line argument for error reporting.
**  Dies on any failure.
*/
static void
create_socket(struct binding *binding, const char *spec)
{
    int fd;
#if defined(SO_REUSEADDR) || defined(IPV6_V6ONLY)
    int flag;
#endif

    /* Create the socket. */
    if (binding->family == AF_INET)
        fd = socket(PF_INET, SOCK_STREAM, IPPROTO_IP);
#ifdef HAVE_INET6
    else if (binding->family == AF_INET6)
        fd = socket(PF_INET6, SOCK_STREAM, IPPROTO_IP);
#endif
    else
        die("unknown protocol family %d in %s", binding->family, spec);
    if (fd < -1)
        sysdie("cannot create socket for %s", spec);

    /* Mark it reusable if possible. */
#ifdef SO_REUSEADDR
    flag = 1;
    if (setsockopt(fd, SOL_SOCKET, SO_REUSEADDR, &flag, sizeof(flag)) < 0)
        sysdie("cannot mark socket reusable for %s", spec);
#endif

    /* Mark it IPv6 only if possible. */
#ifdef IPV6_V6ONLY
    flag = 1;
    if (binding->family == AF_INET6
        && setsockopt(fd, IPPROTO_IPV6, IPV6_V6ONLY, &flag, sizeof(flag)) < 0)
        sysdie("cannot mark socket IPv6 only for %s", spec);
#endif

    /* Fill in the struct. */
    binding->fd = fd;
}


/*
**  Given a file descriptor, attempt to pass it via stdout.  Die on any error.
**  Currently, file descriptor passing is only implemented for systems that
**  use STREAMS.
*/
#ifdef HAVE_STREAMS_SENDFD
static void
send_fd(int fd)
{
    if (isastream(STDOUT_FILENO) != 1)
        die("cannot pass file descriptor: stdout is not a stream");
    if (ioctl(STDOUT_FILENO, I_SENDFD, fd) < 0)
        sysdie("cannot pass file descriptor");
}
#else  /* !HAVE_STREAMS_SENDFD */
__attribute__((__noreturn__)) static void
send_fd(int fd UNUSED)
{
    die("cannot pass file descriptor: STREAMS not supported");
}
#endif /* !HAVE_STREAMS_SENDFD */


int
main(int argc, char *argv[])
{
    uid_t real_uid, uid;
    int i;
    bool done;
    struct binding binding = {0, 0, NULL, 0};
    bool force_sendfd = false;

    /* Set up the error handlers.  Errors go to stderr and to syslog with a
       priority of LOG_CRIT.  This priority level is too high, but it's chosen
       to match innd. */
    openlog("innbind", LOG_CONS, LOG_INN_PROG);
    message_handlers_die(2, message_log_stderr, message_log_syslog_crit);
    message_program_name = "innbind";

    /* If we're running privileged (effective and real UIDs are different),
       convert runasuser to a UID and exit if run by another user.  Don't do
       this if we're not running privileged to make installations that don't
       need privileged ports easier and to make testing easier. */
    real_uid = getuid();
    if (real_uid != geteuid()) {
        get_news_uid_gid(&uid, NULL, true);
        if (real_uid != uid) {
            die("must be run by runasuser (%lu), not %lu", (unsigned long) uid,
                (unsigned long) real_uid);
        }
    }

    /* If the first argument is -p, force creation of the socket and file
       descriptor passing rather than even attempting to bind the socket
       first. */
    if (argc > 1 && strcmp(argv[1], "-p") == 0) {
        force_sendfd = true;
        argc--;
        argv++;
    }
    if (argc < 2)
        die("no addresses specified");

    /* Walk the argument list and try to bind each argument.  For each
       successful bind, print "ok\n" to stdout.  For each bind that fails with
       permission denied, print "no\n" to stdout and then immediately attempt
       to create a new file descriptor and pass it back over stdout. */
    for (i = 1; i < argc; i++) {
        parse_argument(argv[i], &binding);
        done = false;
        if (!force_sendfd)
            done = bind_address(&binding, argv[i]);
        if (done) {
            if (write(STDOUT_FILENO, "ok\n", 3) < 3)
                sysdie("cannot write status");
        } else {
            if (write(STDOUT_FILENO, "no\n", 3) < 3)
                sysdie("cannot write status");
            create_socket(&binding, argv[i]);
            if (!bind_address(&binding, argv[i]))
                sysdie("cannot bind socket for %s", argv[i]);
            send_fd(binding.fd);
        }
    }
    exit(0);
}
